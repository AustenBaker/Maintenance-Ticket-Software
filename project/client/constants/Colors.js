const tintColor = '#2f95dc';

export default {
  tintColor,
  tabIconDefault: '#ccc',
  tabIconSelected: tintColor,
  tabBar: '#000',
  errorBackground: 'red',
  errorText: '#fff',
  warningBackground: '#EAEB5E',
  warningText: '#666804',
  noticeBackground: tintColor,
  noticeText: '#fff',
};
