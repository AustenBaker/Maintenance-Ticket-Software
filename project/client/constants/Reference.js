// REGEX constant 
export const REGEX = {
    // name pattern: up to 32 ASCII alphabetical characters
    NAME: /^[a-zA-Z]{1,32}$/,
    // email pattern:  some(.name)*@site.com
    // where some name & site are any combination of letters
    // and numbers, and the site ends in a 2-3 letter/digit extension
    EMAIL: /^[[\w]+[\.\w]*@[\w]+\.[\w]{2,3}]{6,32}$/,
    // phone pattern:  ###-###-#### where # is a digit 0-9
    PHONE: /^[\d]{3}-[\d]{3}-[\d]{4}$/,
    // password pattern: a-zA-Z0-9_-!@#$%^&*+=
    // requires at least one each: uppercase letter, number, punctuation
    // length of 8-32 characters
    PASSWORD: /^[[\w]*[A-Z]+[\d]+[\-!@#$%\^\*\+=]+]{8,32}$/
}


export const USER_TYPE = ['resident','maintenance','management'];
export const ENTRY_PERMISSION = ['accompanied','notify','any'];
export const PREFERRED_CONTACT = ['email','text'];

