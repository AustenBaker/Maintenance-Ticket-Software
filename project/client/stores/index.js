import { extendObservable } from 'mobx';

export class UserStore {
    constructor() {
        extendObservable(this, {
            loading: true,
            loggedIn: false,
            /* Add new global states here */
        });
    }
}

export class TicketStore {
    constructor() {
        extendObservable(this, {
            /* Add new global states here */
        });
    }
}
