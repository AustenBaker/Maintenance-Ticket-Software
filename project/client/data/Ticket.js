import * as React from 'react';
import { View, CheckBox, Picker, FormInput } from 'react-native';
import * as CONSTANTS from '../constants/Reference';

export default class Ticket extends React.Component{
  constructor(props) {
    super(props);
    this.state = {
      ticket_number: props.ticket_number,
      emergency: props.emergency,
      unit: props.unit,
      first_name: props.first_name,
      last_name: props.last_name,
      status: props.status,
      edit_mode: props.edit_mode,
      ticket_updates: [props.ticket_updates]
    }
  }

  /**
   * Declare default values for Ticket props
   */
  static defaultProps = {
    ticket_number: "99999",
    emergency: true,
    unit: "316",
    first_name: "FirstName",
    last_name: "LastName",
    status: "in-progress",
    edit_mode: "false",
    ticket_updates: []
  }

  //method to create ticket
  createTicket = () => {
	  
  }
  
  
  
  
  /**
   * This method displays tickets
   * emergency level, ticket number, unit number, status
   * display list of tickets
   * this should be called when a ticket from the list is pressed
   */
  displayTicketDetails = () => {
    var content;

    // create <Text> container for emergency data
    var emergency = "";
    if (this.emergency === true){
      emergency = (
        <Text>
          URGENT: This is an emergency level ticket!
        </Text>
      );
    }else{
      emergency = NULL;
    }

    //prepare content
    content = (
      <View>
        <Text>
          Ticket Number:
          {this.state.ticket_number}
          Unit Number:
          {this.state.unit}
          Status:
          {this.state.status}
        </Text>
        {emergency}
		//TODO add edit button
      </View>
    );
    return content;
  }
  
  //display ticket list using flatlist

  /**
   * This method updates ticket information
   * 
   * @param ticket_number Ticket number/id
   * @param emergency Ticket level
   * @param unit Ticket unit number
   * @param status Ticket status
   * 
   * @returns true if updated correctly
   */
  update = (props) => {
    var updated = false;
    var checked = [];
    if('ticket_number' in props){
      checked.push({['ticket_number']: props.ticket_number});
	}
    if('emergency' in props){
	  checked.push({['emergency']: props.emergency});
    }
    if('unit' in props){
	  checked.push({['unit']: props.unit});
    }
    if('status' in props){
	  checked.push({['status']: props.status});
    }

    //TODO: update server state and this.setState
    return updated;
  }

  /**
   * TODO: UPDATE
   * Generates form to display when editing a ticket
   * so far includes:
   * Ticket Number, Unit Number, Emergency/Ticket Level Checkbox,
   * Ticket Status Checkbox,  
   */
  editTicket = () => {
    var content;
    content = (
      <form name = "editTicket">
        <FormLabel>Ticket Number</FormLabel>
        <FormInput>
          onSubmitEditing={new_ticket_number => this.setState(ticket_number, new_ticket_number)}
          defaultValue={this.state.ticket_number}
        </FormInput>
        <FormValidationMessage>{'This field is required.'}</FormValidationMessage>

        <FormLabel>Unit Number</FormLabel>
        <FormInput>
          onSubmitEditing={newUnit => this.setState(unit, newUnit)}
          defaultValue={this.state.unit}
        </FormInput>
        <FormValidationMessage>{'This field is required.'}</FormValidationMessage>

        <FormLabel>Ticket Importance Level:</FormLabel>
        <Picker>
          <Picker
            title='Emergency'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={emergency === true}
            onIconPress={emergency = true}
          />
          <Picker.Item
            label='Normal'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={emergency === false}
            onIconPress={emergency = false}
          />
          <P
            label='Low'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={emergency === false}
            onIconPress={emergency = false}
          />
        </Picker>
        <FormValidationMessage>{'This field is required.'}</FormValidationMessage>

        <FormLabel>Ticket Status:</FormLabel>
        <Picker>
          <Picker.Item
            title='Open'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={status === 'open'}
            onIconPress={status = 'open'}
          />
          <Picker.Item
            title='In-Progress'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={status === 'in-progress'}
            onIconPress={status = 'in-progress'}
          />
          <Picker.Item
            title='Closed'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={status === 'closed'}
            onIconPress={status = 'closed'}
          />
        </Picker>
        <FormValidationMessage>{'This field is required.'}</FormValidationMessage>

      </form>
    );
    return content;
  }
  
  //create a close ticket method
  //method to delete tickets for management

  render = () => {
    var content;
    if(this.state.editMode === false){
      content = this.displayTicket();
    }else{
      content = this.editUser();
    }
    return (content);
  }

}