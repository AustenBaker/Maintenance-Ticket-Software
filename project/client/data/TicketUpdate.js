import * as React from 'react';
import { View, Button, Text, TextInput, Picker, StyleSheet } from 'react-native';
import * as CONSTANTS from '../constants/Reference';
import { colorScheme, userStore } from '../stores/index';

export default class TicketUpdate extends React.Component {
    constructor (props) {
        super(props);
        this.state = {...TicketUpdate.defaultProps};
        for (let key in props) {
            this.state[key] = this.props[key];
        }
        this.toggleEmergency = this.toggleEmergency.bind(this);
        this.updateTicket = this.updateTicket.bind(this);
    }

    static defaultProps = {
        user: 'none',
        timestamp: Date.now(),
        details: 'No changes.',
        edit_mode: false
    }

    /**
     * If update changes emergency ticket status, update ticket
     * emergency status and return boolean value true if ticket
     * status is now set as an Emergency ticket.
     *
     * @return boolean True if ticket is now flagged as an Emergency ticket.
     */
    toggleEmergency = () => {
        return super.toggleEmergency(this);
    }

    updateTicket = () => {
        return super.addUpdate(this);
    }

    makeUpdate = () => {
        let content = (
            <View style={styles.container}>
              <Text style={styles.ticketUpdateTheme}>
                <Text>
                  Ticket Update:
                  {' '}
                </Text>
                <TextInput
                  label='Details'
                  placeholder={'Additional information.'}
                  value={this.state.details}
                  onValueChange={(value) => this.setState({ details: value})}
                >
                </TextInput>
                <Button
                  title='Add Update'
                  label='Update Ticket'
                  onPress={() => {() => this.setState({
                      timestamp: Date.now(),
                      user: userStore.email,
                      edit_mode: false,
                    }), () => {
                        this.updateTicket();
                        // TODO: Redirect to Ticket display
                  }}}
                >
                  <Text>
                    Update Ticket
                  </Text>
                </Button>
                <Button
                  title='Cancel Update'
                  label='Cancel'
                  onPress={() => {alert(`TODO: return to ticket details`), () => {
                      // TODO: unmount aborted ticket update & return to ticket display
                  }}}
                >
                  <Text>
                    Cancel
                  </Text>
                </Button>
              </Text>
            </View>
        );
        return content;
    }

    getUpdate = () => {
        let content = (
            <View style={styles.container}>
                <Text style={styles.ticketUpdateTheme}>
                    <Text>
                        Updated:
                        {'  '}
                        {new Date(this.state.timestamp).toString()}
                    </Text>
                    <Text>
                        By:
                        {'  '}
                        {this.state.user}
                    </Text>
                </Text>
                <Text style={styles.ticketUpdateTheme}>
                    {this.state.details}
                </Text>
            </View>
        );
        return content;
    }

    render () {
        let content = '';
        if (this.state.edit_mode) {
            content = this.makeUpdate();
        } else {
            content = this.getUpdate();
        }
        return content;
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'column',
        alignContent: 'stretch',
        alignItems: 'flex-start',
        alignSelf: 'center',
    },
    ticketUpdateTheme: {
        fontSize: 16,
    }
});
