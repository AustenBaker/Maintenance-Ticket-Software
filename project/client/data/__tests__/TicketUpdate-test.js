import * as React from 'react';
import { Text, TextInput } from 'react-native';
import renderer from 'react-test-renderer';
import { fireEvent, render, wait, getByPlaceholderText, getByDisplayValue, getByRole, getByLabelText, getByTestId, getByTitle, getByText } from '@testing-library/react-native';
import * as JestNative from '@testing-library/jest-native';
import * as CONSTANTS from '../../constants/Reference';
import Ticket from '../Ticket';
import TicketUpdate from '../TicketUpdate';
import { userStore, ticketStore, colorScheme } from '../../stores/index';


expect.extend({...JestNative});

jest.disableAutomock();

jest.mock('expo', () => ({
    AppLoading: 'AppLoading',
}));

describe(`TicketUpdate`, () => {
    jest.useFakeTimers();

    beforeEach(() => {
      // TODO: test setup
    });

    afterEach(() => {
      // TODO: test cleanup
    });


    it(`renders the ticket update`, () => {
      const tree = renderer.create(<TicketUpdate timestamp={0}/>).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it(`renders the ticket update creation screen`, () => {
      const tree = renderer.create(<TicketUpdate edit_mode timestamp={0}/>).toJSON();
      expect(tree).toMatchSnapshot();
    });
});
