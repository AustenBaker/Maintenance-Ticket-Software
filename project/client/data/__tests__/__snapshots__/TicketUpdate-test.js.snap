// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketUpdate renders the ticket update 1`] = `
<View
  style={
    Object {
      "alignContent": "stretch",
      "alignItems": "flex-start",
      "alignSelf": "center",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 16,
      }
    }
  >
    <Text>
      Updated:
        
      12/31/1969 6:00:00 PM
    </Text>
    <Text>
      By:
        
      none
    </Text>
  </Text>
  <Text
    style={
      Object {
        "fontSize": 16,
      }
    }
  >
    No changes.
  </Text>
</View>
`;

exports[`TicketUpdate renders the ticket update creation screen 1`] = `
<View
  style={
    Object {
      "alignContent": "stretch",
      "alignItems": "flex-start",
      "alignSelf": "center",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 16,
      }
    }
  >
    <Text>
      Ticket Update:
       
    </Text>
    <TextInput
      allowFontScaling={true}
      label="Details"
      onValueChange={[Function]}
      placeholder="Additional information."
      rejectResponderTermination={true}
      underlineColorAndroid="transparent"
      value="No changes."
    />
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Add Update
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Cancel Update
        </Text>
      </View>
    </View>
  </Text>
</View>
`;
